go/tasks/plugins/k8s/ray/ray.go

```diff
@@ -0,0 +1,353 @@
+		headNodeRayStartParams = rayJob.RayCluster.HeadGroupSpec.RayStartParams
+	}
+	headNodeRayStartParams["include-dashboard"] = "true"
+	headNodeRayStartParams["node-ip-address"] = "$MY_POD_IP"
+	headNodeRayStartParams["dashboard-host"] = "0.0.0.0"
+
+	enableIngress := true
+	rayClusterSpec := rayv1alpha1.RayClusterSpec{
+		HeadGroupSpec: rayv1alpha1.HeadGroupSpec{
+			Template:       buildHeadPodTemplate(podSpec),
+			ServiceType:    v1.ServiceType(GetConfig().ServiceType),
+			Replicas:       &headReplicas,
+			EnableIngress:  &enableIngress,
+			RayStartParams: headNodeRayStartParams,
+		},
+		WorkerGroupSpecs: []rayv1alpha1.WorkerGroupSpec{},
+	}
+
+	for _, spec := range rayJob.RayCluster.WorkerGroupSpec {
+		workerPodTemplate := buildWorkerPodTemplate(podSpec)
+
+		minReplicas := spec.Replicas
+		maxReplicas := spec.Replicas
+		if spec.MinReplicas != 0 {
+			minReplicas = spec.MinReplicas
+		}
+		if spec.MaxReplicas != 0 {
+			maxReplicas = spec.MaxReplicas
+		}
+
+		workerNodeRayStartParams := make(map[string]string)
+		if spec.RayStartParams != nil {
+			workerNodeRayStartParams = spec.RayStartParams
+		}
+		workerNodeRayStartParams["node-ip-address"] = "$MY_POD_IP"
+
+		workerNodeSpec := rayv1alpha1.WorkerGroupSpec{
+			GroupName:      spec.GroupName,
+			MinReplicas:    &minReplicas,
+			MaxReplicas:    &maxReplicas,
+			Replicas:       &spec.Replicas,
+			RayStartParams: workerNodeRayStartParams,
+			Template:       workerPodTemplate,
+		}
+
+		rayClusterSpec.WorkerGroupSpecs = append(rayClusterSpec.WorkerGroupSpecs, workerNodeSpec)
+	}
+
+	serviceAccountName := flytek8s.GetServiceAccountNameFromTaskExecutionMetadata(taskCtx.TaskExecutionMetadata())
+
+	rayClusterSpec.HeadGroupSpec.Template.Spec.ServiceAccountName = serviceAccountName
+	for index := range rayClusterSpec.WorkerGroupSpecs {
+		rayClusterSpec.WorkerGroupSpecs[index].Template.Spec.ServiceAccountName = serviceAccountName
+	}
+
+	jobSpec := rayv1alpha1.RayJobSpec{
+		RayClusterSpec:           rayClusterSpec,
+		Entrypoint:               strings.Join(podSpec.Containers[0].Args, " "),
+		ShutdownAfterJobFinishes: GetConfig().ShutdownAfterJobFinishes,
+		TTLSecondsAfterFinished:  &GetConfig().TTLSecondsAfterFinished,
+		RuntimeEnv:               rayJob.RuntimeEnv,
+		ClusterSelector:          GetConfig().ClusterSelector,
+	}
+
+	rayJobObject := rayv1alpha1.RayJob{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       KindRayJob,
+			APIVersion: rayv1alpha1.SchemeGroupVersion.String(),
+		},
+		Spec: jobSpec,
+	}
+
+	return &rayJobObject, nil
+}
+
+func buildHeadPodTemplate(podSpec *v1.PodSpec) v1.PodTemplateSpec {
+	// Some configs are copy from  https://github.com/ray-project/kuberay/blob/b72e6bdcd9b8c77a9dc6b5da8560910f3a0c3ffd/apiserver/pkg/util/cluster.go#L97
+	// They should always be the same, so we could hard code here.
+	primaryContainer := &v1.Container{Name: "ray-head", Image: podSpec.Containers[0].Image}
+	primaryContainer.Resources = podSpec.Containers[0].Resources
```

Comment:
I think we need to be VERY careful here. I'm not sure how containers are created here, but it is worrisome to use indices. Is there a scenario where multiple containers are defined and we update the wrong one?

Reply:
Used `ToK8sContainer` instead now.